### PR TITLE
Current buffer last buffer

### PR DESCRIPTION
**Pull request by Ambrevar**
Description in the various commits.

### CODE REVIEW

File: source/configuration.lisp [hunk 1]
```diff
@@ -24,6 +24,28 @@ from a binary) then any condition is logged instead of triggering the debugger."
           (log:error "In ~a: ~a" f c)
           nil))))
 
+(defparameter %buffer nil)              ; TODO: Make a monad?
```
Comment (jmercouris) [inline source/configuration.lisp:27]: I don't understand what a monad is still :-D, can you please explain how it would function and what the benefits are?

File: source/search-engine.lisp [hunk 2]
```diff
@@ -37,7 +37,8 @@ the empty string."))
           (search-url engine)))
 
 (defun bookmark-search-engines (&optional (bookmarks (get-data (bookmarks-path
-                                                                (current-buffer)))))
+                                                                (or (current-buffer)
```
Comment (jmercouris) [inline source/search-engine.lisp:40]: from my understanding, =(current-buffer)= will always return a buffer because it will resort to =(last-active-buffer)= which will just go through all buffers. Therefore, will there ever be an instance of a user-buffer created here?

**Comment by jmercouris** [comment]
I like it! Now we can be sure that (current-buffer) will return EXACTLY what we expect within a command body! :-)

Very neatly done too!

File: source/configuration.lisp (hunk 1 above)
Comment (Ambrevar) [reply source/configuration.lisp:27]: See https://github.com/atlas-engineer/nyxt/issues/868#issuecomment-673024703
Not that this comment is from before, nothing has changed here.

File: source/search-engine.lisp (hunk 2 above)
Comment (Ambrevar) [reply source/search-engine.lisp:40]: There is, when you instantiate a browser but you didn't start it!  It's a corner case but it's an important one for tests!

File: source/configuration.lisp (hunk 1 above)
Comment (jmercouris) [reply source/configuration.lisp:27]: I still don't see why we need a monad when we can use a closure

File: source/search-engine.lisp (hunk 2 above)
Comment (jmercouris) [reply source/search-engine.lisp:40]: Ah, OK, in that case there will be no buffers!

Comment (jmercouris) [reply source/search-engine.lisp:40]: We can make a user buffer object without starting the browser, but we can't actually interact with it, so this function won't actually work, will it?

File: source/search-engine.lisp [hunk 3]
```diff
@@ -52,7 +53,8 @@ the empty string."))
 (defun all-search-engines ()
   "Return the `search-engines' from the `browser' instance plus those in
 bookmarks."
-  (sera:and-let* ((buffer (current-buffer)))
+  (sera:and-let* ((buffer (or (current-buffer)
```
Comment (aartaka) [inline source/search-engine.lisp:56]: We probably don't need `sera:and-let*` here anymore, because there always will be an instance of `user-buffer` bound to `buffer`.

Comment (Ambrevar) [reply source/search-engine.lisp:56]: Good catch!

File: source/search-engine.lisp (hunk 2 above)
Comment (Ambrevar) [reply source/search-engine.lisp:40]: It will, because bookmarks-path will return the default value.
This technique of making a buffer to access the defaults is used a couple of times, e.g. in the tutorial / manual.

File: source/configuration.lisp (hunk 1 above)
Comment (Ambrevar) [reply source/configuration.lisp:27]: Because we need to reset %buffer when we change buffer within a `with-current-buffer`.  This is stateful, global and thus brittle.

A monad would give us full control over the side effects.

Comment (Ambrevar) [reply source/configuration.lisp:27]: Note: currently we _do_ reset %buffer in the only place (I think) that changes the current buffer.  So at the moment our code should be correct.

File: source/search-engine.lisp (hunk 2 above)
Comment (jmercouris) [reply source/search-engine.lisp:40]: Aha! I see!

**Comment by jmercouris** [comment]
Please merge whenever you are ready!

File: source/configuration.lisp (hunk 1 above)
Comment (jmercouris) [reply source/configuration.lisp:27]: I think the reason we have this problem is because %buffer is serving two purposes:

- result of active buffer
- place to save current-buffer when wrapping in with-current-buffer

Comment (jmercouris) [reply source/configuration.lisp:27]: therefore we have to update %buffer when we switch buffer within a within-current-buffer form, whereas we shouldn't have to, they should be discrete (I think)

Comment (Ambrevar) [reply source/configuration.lisp:27]: The goal of `with-current-buffer` is to ensure that `(current-buffer)` returns a given buffer.

%buffer is used in only one spot: `current-buffer`.

What is the other purpose that you see?

Comment (jmercouris) [reply source/configuration.lisp:27]: but as we just said, if we change the current buffer then %buffer is no longer valid, and (current-buffer) must return a different buffer. there are two concepts here

(with-current-buffer ...) should have a symbol/variable for the buffer that is the current buffer at the time of creation, and (current-buffer) should always just return the current buffer.

am I making sense? :-D I feel like it is a bit hard to explain the subtle difference here.

Comment (Ambrevar) [reply source/configuration.lisp:27]: OK, but this is not what we want.  We effectively want to locally override the return value of (current-buffer), and we don't want to resort to another function/variable to get the possibly-modified current buffer.

In other words, we don't want to write code that knows in advance whether it should use `current-buffer` or `buffer-from-with-current-buffer` (I made this last one up).

Comment (jmercouris) [reply source/configuration.lisp:27]: well, in that case, then we must deal with the state :-D

Comment (Ambrevar) [reply source/configuration.lisp:27]: This is exactly what monads allow us to do cleanly ;)

**Comment by Ambrevar** [comment]
OK to merge then?

**Comment by jmercouris** [comment]
OK to merge :-)

**Comment by Ambrevar** [comment]
Done in ccde4fbab58e94092a0c2d44dfcda8e9e33b914c.

**Comment by jmercouris** [comment]
Thank you monsieur!
